### PR TITLE
Hw1: add FastAPI for interacting with mail client

### PR DESCRIPTION
**Pull request by ysjprojects**
# Pull Request

## Summary
- Added `mail_client_service` with FastAPI application
- Added unit, integration, e2e tests for `mail_client_api`,  `mail_client_service`, `mail_client_adapter`, `gmail_client_impl`

## Related Issues
- Closes #
- Relates to #

## Change Type
- [ ] 🐛 Bug fix
- [x] ✨ Feature / enhancement
- [ ] 💥 Breaking change
- [ ] 📚 Documentation
- [ ] 🔧 Refactor / cleanup
- [ ] ⚡ Performance
- [x] 🧪 Test improvement
- [ ] 🔨 Build / tooling

## Impacted Areas
- [x] `mail_client_api`
- [x] `gmail_client_impl`
- [ ] Documentation
- [x] Tests
- [ ] Tooling / CI
- [ ] Other (describe in summary)

### Commands executed
```bash
# Example (edit as needed)
uv run pytest src/ tests/
uv run mypy src tests
uv run ruff check .
```

### Results
- [x] Tests pass
- [x] Type checks clean
- [x] Linting clean
- [ ] Manual verification (if applicable)

## Quality Checklist
- [ ] Contract boundaries respected; abstractions unchanged unless noted
- [ ] Backward compatibility confirmed or migration documented
- [ ] Security / secrets handled correctly
- [ ] Performance impact acceptable
- [ ] Docs updated for user-facing changes

## Breaking Changes (skip if none)
1. Impact:
2. Migration / mitigation:

## Notes for Reviewers
Call out tricky areas, follow-ups, or validation details.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -2,11 +2,17 @@
 
 # ta-assignment/main.py
 
+# Import the protocols first
+# --- TRIGGER DEPENDENCY INJECTION ---
+# By importing the implementation packages, their __init__.py files
+# run and override the factory functions in the protocol packages.
+import gmail_client_impl
+import gmail_message_impl
```
Comment (AlperMumcular) [inline main.py:10]: There is no package named "gmail_message_impl" in your repo.

File: pyproject.toml [hunk 2]
```diff
@@ -24,6 +24,9 @@ dev = [
 members = [
   "src/mail_client_api",
   "src/gmail_client_impl",
+  "src/message",
+  "src/gmail_message_impl",
```
Comment (AlperMumcular) [inline pyproject.toml:28]: Line 27 and 28 - Similar issue with main. There is no folder under src named "message" nor "gmail_message_impl".

File: src/mail_client_service/README.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
+# mail-client-service
```
Comment (AlperMumcular) [inline src/mail_client_service/README.md:1]: Structure needs improvement (Check review body message for readme.mds)

File: src/mail_client_service/mail-client-service-client/README.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
+# mail-client-service-client
```
Comment (AlperMumcular) [inline src/mail_client_service/mail-client-service-client/README.md:1]: Structure needs improvement (Check review body message for readme.mds)

File: src/mail_client_service/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+requires = ["hatchling"]
+build-backend = "hatchling.build"
+
+[tool.ruff]
+line-length = 100
+target-version = "py311"
+extend = "../../pyproject.toml"
+
+[tool.ruff.lint]
+ignore = [
```
Comment (AlperMumcular) [inline src/mail_client_service/pyproject.toml:30]: The [tool.ruff.lint] ignore list disables several rules, including some related to code complexity (C901, PLR0912, PLR0915) and security (S101, S105). Consider if these rules need to be ignored globally, or if they could be managed locally in specific cases. Disabling security-related checks may allow issues to slip through.

Why ignoring blank line whitespaces W293? 

You may add a short rationale in the docs or comments for future contributors for these ignores.

File: src/mail_client_service/src/mail_client_service/app.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+                "from": msg.from_,
+                "to": msg.to,
+                "date": msg.date,
+                "subject": msg.subject,
+                "body": msg.body
+            }
+            for msg in messages
+        ]
+    except Exception as e:
+        raise HTTPException(status_code=500, detail="Internal server error") from e
```
Comment (AlperMumcular) [inline src/mail_client_service/src/mail_client_service/app.py:75]: All exceptions in the get_message, list_messages, mark_as_read, and delete_message endpoints are mapped to either a 404 or 500 error, regardless of the actual error cause. This may unintentionally hide issues unrelated to "not found" (e.g., connection errors, invalid input, or unexpected failures).
Suggestion:
Catch specific exceptions for "not found" (e.g., a custom MessageNotFoundError) and use 404.
For other errors, return a 500 or a more appropriate error code.

File: src/mail_client_service/src/mail_client_service/app.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+from pydantic import BaseModel, Field
+from mail_client_api.test_client import TestClient as Client
+
+app = FastAPI(
+    title="Mail Client Service",
+    description="A REST API service that implements mail client operations",
+    version="1.0.0",
+)
+
+def get_client():
```
Comment (AlperMumcular) [inline src/mail_client_service/src/mail_client_service/app.py:17]: get_client is defined but not used. Remove unused functions if not needed.


Both get_client and get_mail_client return a new Client(). You only use get_mail_client as a dependency; consider removing get_client to avoid confusion.

File: src/mail_client_service_impl/pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+requires = ["hatchling"]
+build-backend = "hatchling.build"
+
+[tool.ruff]
+line-length = 100
+target-version = "py311"
+extend = "../../pyproject.toml"
+
+[tool.ruff.lint]
+ignore = [
```
Comment (AlperMumcular) [inline src/mail_client_service_impl/pyproject.toml:22]: Similar issue

File: src/mail_client_service/tests/test_app.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    assert data[0] == {
+        "id": "1",
+        "from": "sender@example.com",
+        "to": "recipient@example.com",
+        "date": "2025-10-03",
+        "subject": "First Message",
+        "body": "Test body for message 1"
+    }
+
+def test_list_messages_with_max_results(mock_mail_client):
```
Comment (AlperMumcular) [inline src/mail_client_service/tests/test_app.py:62]: For test_list_messages_with_max_results, consider adding tests for boundary cases (e.g., max_results greater than available messages, max_results negative, very large values).

File: src/mail_client_service/tests/test_app.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
+"""Unit tests for the FastAPI mail client service.
+
+These tests verify that the FastAPI endpoints handle requests and responses correctly,
+using a mocked mail client to isolate the tests from the actual implementation.
+"""
+
+from datetime import datetime
+from fastapi.testclient import TestClient
+from unittest.mock import patch, MagicMock
+from dataclasses import dataclass
```
Comment (AlperMumcular) [inline src/mail_client_service/tests/test_app.py:10]: Unused Imports:
The import for dataclasses.dataclass appears unused.

File: tests/e2e/test_mail_client_service_e2e.py [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+import os
+import pytest
+import uvicorn
+import multiprocessing
+import time
+
+from mail_client_service_impl import MailClientServiceImpl
+from message import Message
+import gmail_client_impl  # Register Gmail implementation
+import gmail_message_impl  # Register Gmail message implementation
```
Comment (AlperMumcular) [inline tests/e2e/test_mail_client_service_e2e.py:16]: No package named "gmail_message_impl"

File: tests/e2e/test_mail_client_service_e2e.py [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+            port=SERVICE_PORT,
+            log_level="error"
+        )
+
+    # Start service in a separate process
+    process = multiprocessing.Process(target=run_service)
+    process.start()
+    
+    # Wait for service to start
+    time.sleep(2)
```
Comment (AlperMumcular) [inline tests/e2e/test_mail_client_service_e2e.py:40]: Replace the fixed time.sleep(2) delay with a retry-based readiness check to ensure the service is actually available before running tests.

File: tests/integration/test_mail_client_service.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+            log_level="error"
+        )
+
+    # Start service in a separate process
+    process = multiprocessing.Process(target=run_service)
+    process.start()
+    
+    # Wait for service to start
+    import time
+    time.sleep(2)
```
Comment (AlperMumcular) [inline tests/integration/test_mail_client_service.py:47]: Replace the fixed time.sleep(2) delay with a retry-based readiness check to ensure the service is actually available before running tests.

**Review by AlperMumcular — COMMENTED** [review]
Overall great job! I have a few feedbacks/questions:

- Why did you modify src_mail_client api? It was using ABC and you need to make "Talk a little bit about the differences between this and Protocol". You've already explained this in contributing.md but now this package became Protocol.
- If you modify src_mail_client package, then you need to write tests for it as well. My suggestion is keep it like TAs, just add your new components based on TAs architecture.
- README.mds belonging to packages should be in their package folder and at minimum, it should include overview, scope, exposed interfaces, usage pattern, and component dependencies. Keep examples using absolute imports. (Stated in mkdocs)
- Too many ignore ruff lints, which may cause safety issues.

Also before merging:
- Make sure you pass all checks, satisfy coverage percentage. We can't see check results.

File: src/mail_client_api/src/mail_client_api/__init__.py [hunk 14]
```diff
@@ -1,7 +1,98 @@
...
+    Args:
+        interactive (bool): If True, the client may prompt for user input
+        during initialization (e.g., for OAuth2 flow). If False, it will
+        use environment variables or other non-interactive methods.
+
+    Returns:
+        Client: A concrete mail client instance.
+    """
+    from mail_client_api.test_client import TestClient
+    return TestClient()
```
Comment (ander9991) [inline src/mail_client_api/src/mail_client_api/__init__.py:98]: To my understanding, this defeats the dependency injection.

```
def get_client(interactive: bool = False) -> Client:
    from mail_client_api.test_client import TestClient
    return TestClient()
```

When ```gmail_client_impl``` is imported to override the interface, it still returns TestClient. It no longer works as described in CONZTRIBUTING.md (lines 22-30).

Perhaps, you need to change it back to ```raise NotImplementedError``` to match the original design, or specify why TestClient is the default

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+Interfaces are implemented using Python's `abc` module, which allows for the definition of abstract base classes. Implementations in `gmail_client_impl` and `gmail_message_impl` inherit from these ABCs and provide concrete logic. This approach ensures that all required methods are implemented and enables type checking.
+
+Unlike `Protocol` from the `typing` module, which allows for structural subtyping (duck typing), ABCs enforce nominal subtyping, requiring explicit inheritance. Protocols are more flexible for static type checking, but ABCs provide runtime enforcement and can include abstract methods.
+
+### Dependency Injection
+The project uses dependency injection to decouple interface definitions from their implementations. In `main.py` (root), line 8, the injection occurs:
+
+```python
+import gmail_client_impl  # gmail takes priority
+import mail_client_api
```
Comment (ander9991) [inline CONTRIBUTING.md:27]: Not a big deal, but the actual dependency injection is on lines 9-10 in ```main.py```

File: src/mail_client_api/src/mail_client_api/test_client.py [hunk 16]
```diff
@@ -0,0 +1,54 @@
+"""Test implementation of the mail client for testing."""
+
+from collections.abc import Iterator
+from datetime import datetime
```
Comment (ander9991) [inline src/mail_client_api/src/mail_client_api/test_client.py:4]: Unused import

**Review by ander9991 — COMMENTED** [review]
Overall, nicely done. Just noticed a few inconsistencies outside of what Alper mentioned. It also looks like your CircleCI fail.

File: main.py (hunk 1 above)
Comment (koitu) [reply main.py:10]: In an older version of the TA repo there did exist "gmail_message_impl" but it has since been removed

**Review by koitu — COMMENTED** [review]
Great job overall! I just want to comment on a couple of things:
- make sure to update the project README.md your changes
- it looks like the linting and formatting is not passing in CircleCI (I recommend using [pre-commit hook](https://github.com/astral-sh/ruff-pre-commit))

File: tests/integration/test_mail_client_service.py [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+    # Call through service
+    messages = list(service_client.get_messages(max_results=2))
+
+    # Verify results
+    assert len(messages) == 2
+    assert messages[0].id == "1"
+    assert messages[0].subject == "First Message"
+    assert messages[1].id == "2"
+    assert messages[1].subject == "Second Message"
+
```
Comment (koitu) [inline tests/integration/test_mail_client_service.py:89]: I think that this is missing a pytest marker

File: src/mail_client_service/pyproject.toml (hunk 5 above)
Comment (koitu) [reply src/mail_client_service/pyproject.toml:30]: I would recommend against too much global lint ignores since they erode the point of the linter in the first place. If there is a specific instance where you want to override it, you can use `# ruff: noqa: <RULE>` instead. See [ruff error suppression](https://docs.astral.sh/ruff/linter/#error-suppression) for more details.

File: src/mail_client_service/openapi.json [hunk 18]
```diff
@@ -0,0 +1 @@
...,"to":{"type":"string","title":"To","description":"Recipient's email address"},"date":{"type":"string","title":"Date","description":"Date the message was sent"},"subject":{"type":"string","title":"Subject","description":"Subject line of the message"},"body":{"type":"string","title":"Body","description":"Body content of the message","default":""}},"type":"object","required":["id","from","to","date","subject"],"title":"MessageResponse","description":"Represents an email message response.","example":{"body":"This is a test message body","date":"2025-10-03","from":"sender@example.com","id":"msg_123","subject":"Test Message","to":"recipient@example.com"}},"SuccessResponse":{"properties":{"success":{"type":"boolean","title":"Success","description":"Whether the operation was successful"}},"type":"object","required":["success"],"title":"SuccessResponse","description":"Standard success response."},"ValidationError":{"properties":{"loc":{"items":{"anyOf":[{"type":"string"},{"type":"integer"}]},"type":"array","title":"Location"},"msg":{"type":"string","title":"Message"},"type":{"type":"string","title":"Error Type"}},"type":"object","required":["loc","msg","type"],"title":"ValidationError"}}}}
```
Comment (koitu) [inline src/mail_client_service/openapi.json:1]: I don't think this file should be added to the repo, as it is dynamically generated when starting the fastAPI server. Rather than needing to remember to update this whenever the API is modified, there is less chance for error if it generated each time.

File: src/mail_client_api/src/mail_client_api/test_client.py
Comment (adithyab-20) [inline src/mail_client_api/src/mail_client_api/test_client.py:1]: Why are we creating this? The interface is defined in mail_client_api.Client.

File: src/mail_client_service/src/mail_client_service/app.py [hunk 19]
```diff
@@ -0,0 +1,125 @@
+"""FastAPI service implementation for the mail client.
+
+This module provides a REST API service that implements the mail client operations.
+"""
+
+from typing import List
+from fastapi import FastAPI, Depends, HTTPException, Query, Path
+from pydantic import BaseModel, Field
+from mail_client_api.test_client import TestClient as Client
```
Comment (adithyab-20) [inline src/mail_client_service/src/mail_client_service/app.py:9]: This is not correct -- you have to import the interface from mail_client_api.Client and then the implementation from gmail_client_impl. Then use the get_client() method to to do dependency injection.

File: src/mail_client_service/openapi.json (hunk 18 above)
Comment (adithyab-20) [reply src/mail_client_service/openapi.json:1]: Yep, this should not be part of the repo.

File: src/mail_client_service/pyproject.toml [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+[build-system]
+requires = ["hatchling"]
+build-backend = "hatchling.build"
+
+[tool.ruff]
+line-length = 100
+target-version = "py311"
+extend = "../../pyproject.toml"
+
+[tool.ruff.lint]
```
Comment (adithyab-20) [inline src/mail_client_service/pyproject.toml:29]: Use the root pyproject.toml to define your ruff and mypy rules and config. You should not add them to the component pyproject.toml's

File: src/mail_client_service/pyproject.toml (hunk 5 above)
Comment (adithyab-20) [reply src/mail_client_service/pyproject.toml:30]: Yes, good feedback. As a general rule, you should allow all checks and only disable them if you really need to and document why.

File: src/mail_client_api/src/mail_client_api/__init__.py (hunk 14 above)
Comment (adithyab-20) [reply src/mail_client_api/src/mail_client_api/__init__.py:98]: Yeah exactly, the interface is already defined. Why are we doing this?

File: src/mail_client_service_impl/src/mail_client_service_impl/__init__.py [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+
+This module provides a concrete implementation of the Client protocol
+that uses the REST API service to perform mail client operations.
+"""
+
+from typing import Iterator
+from mail_client_api import Client
+from mail_client_service_client import Client as ServiceClient
+from mail_client_service_client.models import MessageResponse
+from message import Message
```
Comment (adithyab-20) [inline src/mail_client_service_impl/src/mail_client_service_impl/__init__.py:11]: Don't use relative imports, always absolute imports.

File: tests/e2e/test_mail_client_service_e2e.py (hunk 11 above)
Comment (adithyab-20) [reply tests/e2e/test_mail_client_service_e2e.py:16]: Yes!

**Comment by austinhuang0131** [comment]
Superseded by #4 .
